### ISSUE TITLE
How to generate lidar point cloud?

### ISSUE THREAD

**Issue by Sibozhu** [Tom-Huang/carla_data_collection#4]
Dear @Tom-Huang ,

Thanks for the previous help in compiling your code.

In the README, you said `You can modify the code to generate the lidar point cloud`, can you actually point out which of those lines to be uncomment? I found there's a function called `lidar_saver` but it isn't being called anywhere.

Once I can save and generate the correct lidar point cloud with semantic label, I'll clean up the code and submit a pull request to you, so people in the future can benefit as well.

Best,
Sibo

**Comment by Sibozhu** [comment]
Also, I found code for `BufferedImageSaver` from somewhere else, but couldn't find `BufferedLidarSaver`, is this a necessary function for generating lidar point cloud?

**Comment by maskjp** [comment]
Hi, @Sibozhu! Can I ask a question about the recompile? Except for copy the uasset files in the folder, do we need to change anything in the project file for recompiling Carla? 
Because after I recompile Carla, I still can't use the normal camera in the file. And I opened the unreal project file. I found the project didn't load the new uasset file. 

Thank you!

**Comment by Sibozhu** [comment]
@maskjp ,have you tried [issue #2](https://github.com/Tom-Huang/carla_data_collection/issues/2) ?

**Comment by maskjp** [comment]
> @maskjp ,have you tried [issue #2](https://github.com/Tom-Huang/carla_data_collection/issues/2) ?

Hi, @Sibozhu! Thank you for your reply. Yes, I deleted the build folder and run the make clean command, but the project still can't load the normal camera and others. I only want to use the lidar semantic so I removed code related, to these sensors. But I found the code only can save the image. And no code saves the lidar data.

Btw, saving images seems not to require BufferedImageSaver.

**Comment by Sibozhu** [comment]
@maskjp 
You are absolutely right, using BufferedImageSaver will save the image into .ply format, which is the ideal format that we want to save lidar point cloud into. I believe the BufferedLiDARSaver is code modified from BufferedImageSave.

`project can't load normal cameras` meaning you still can't save images? 

If you are looking for a codebase that compiles and can generate rgb and depth image (semantic segmentation image still need to be post-processed, which we can make it generated in real-time in the future), you can try [my fork](https://github.com/Sibozhu/carla_data_collection). We can work on making the codebase clean and functioning together and submit pull request to @Tom-Huang in the end. (also added autopilot feature so the simulated agent can have some driving footage)

![demo](https://user-images.githubusercontent.com/22118253/87207915-1d189c80-c2db-11ea-8f69-1eadd5f41f94.gif)

**Comment by maskjp** [comment]
> @maskjp
> You are absolutely right, using BufferedImageSaver will save the image into .ply format, which is the ideal format that we want to save lidar point cloud into. I believe the BufferedLiDARSaver is code modified from BufferedImageSave.
> 
> `project can't load normal cameras` meaning you still can't save images?
> 
> If you are looking for a codebase that compiles and can generate rgb and depth image (semantic segmentation image still need to be post-processed, which we can make it generated in real-time in the future), you can try [my fork](https://github.com/Sibozhu/carla_data_collection). We can work on making the codebase clean and functioning together and submit pull request to @Tom-Huang in the end. (also added autopilot feature so the simulated agent can have some driving footage)
> 
> ![demo](https://user-images.githubusercontent.com/22118253/87207915-1d189c80-c2db-11ea-8f69-1eadd5f41f94.gif)

@Sibozhu Thank you for your reply!
I mean, the Unreal engine project can't load newly added uassets even after I deleted all the intermediate files and recompile the Carla.
But I am currently interested in getting labeled lidar data, so I may don't need the four new uassets. 

BTW, to save lidar point cloud, I found [a sample code in carla](https://github.com/carla-simulator/carla/blob/f324c73b1012aae74da29a9a892f3954a128ec4c/PythonAPI/examples/automatic_control.py#L641) and the [save_to_disk](https://carla.readthedocs.io/en/latest/python_api/#carlalidarmeasurement) method.

**Comment by Sibozhu** [comment]
> > @maskjp
> > You are absolutely right, using BufferedImageSaver will save the image into .ply format, which is the ideal format that we want to save lidar point cloud into. I believe the BufferedLiDARSaver is code modified from BufferedImageSave.
> > `project can't load normal cameras` meaning you still can't save images?
> > If you are looking for a codebase that compiles and can generate rgb and depth image (semantic segmentation image still need to be post-processed, which we can make it generated in real-time in the future), you can try [my fork](https://github.com/Sibozhu/carla_data_collection). We can work on making the codebase clean and functioning together and submit pull request to @Tom-Huang in the end. (also added autopilot feature so the simulated agent can have some driving footage)
> > ![demo](https://user-images.githubusercontent.com/22118253/87207915-1d189c80-c2db-11ea-8f69-1eadd5f41f94.gif)
> 
> @Sibozhu Thank you for your reply!
> I mean, the Unreal engine project can't load newly added uassets even after I deleted all the intermediate files and recompile the Carla.
> But I am currently interested in getting labeled lidar data, so I may don't need the four new uassets.
> 
> BTW, to save lidar point cloud, I found [a sample code in carla](https://github.com/carla-simulator/carla/blob/f324c73b1012aae74da29a9a892f3954a128ec4c/PythonAPI/examples/automatic_control.py#L641) and the [save_to_disk](https://carla.readthedocs.io/en/latest/python_api/#carlalidarmeasurement) method.

Thanks! @maskjp This is definitely helpful!
